### ISSUE TITLE
Support method/function overloads

### ISSUE THREAD

**Issue by nex3** [dart-lang/language#1122]
This has been discussed periodically both in the issue tracker and in person, but I don't think there's a tracking issue yet. (**EDIT**: Here is the original issue from 2011 - https://github.com/dart-lang/sdk/issues/49).

Now that we're moving to a sound type system, we have the ability to overload methods—that is, to choose which of a set of methods is called based on which arguments are passed and what their types are. This is particularly useful when the type signature of a method varies based on which arguments are or are not supplied.

**Comment by Terrain2** [comment]
How about making sure every overload has a unique name? This would fix the issue with tearoffs

With [matanlurey's example for a dice roll](https://github.com/dart-lang/language/issues/1122#issuecomment-403124250)

```dart
abstract class DiceRoller {
  List<int> roll.parse(String expression);
  List<int> roll.many(int amount, int sides);
  List<int> roll.from(DiceRoll roll);
}
```

And then don't allow tearing off the method, only the full overload's name. The actual overload if accessed by a dynamic getter would be some automatically generated `Overload` class or whatever, and then it's not at all callable directly, only through the different overloads. When writing code with proper types, you can either do the same and call the full overload name, or if you just called the method name, it would be replaced with a specific overload at compile time.

```dart
void main() {
  var roller = DiceRollerImpl();

  var parsed = roller.roll("4d6"); // Replaced with roller.roll.parse("4d6"); at compile time
  var parsed2 = roller.roll.parse("4d6"); // Identical to the above, except slightly more verbose
  var many = roller.roll(4, 6); // Replaced with roller.roll.many(4, 6); at compile time

  var roll = roller.roll; // Error: Cannot tear off an overloaded method directly
  List<int> Function(String) roll2 = roller.roll; // Error: Not even unambiguos tearoffs

  var rollParse = roller.roll.parse; // List<int> Function(String)
  var rollMany = roller.roll.many; // List<int> Function(int, int)

  dynamic roller2 = roller;
  var parsedError = roller2.roll("4d6"); // Error: overload cannot be invoked dynamically
  var dynamicRoll = roller2.roll; // This is fine, simply has a couple methods on it
  var dynamicParsed = dynamicRoll.parse("4d6"); // Rolls fine at runtime
  assert(rollParse == dynamicRoll.parse); // Same method

  
}
```

Alternatively, overloads could be compiled with unique names instead of nesting under a single "overload object", but then getting that property dynamically might be a bit hard (though, to be honest, how useful is that anyways? tearoffs would still work the same given you have the correct static type)

Since each one *does* have a unique name and won't ever become a single object with multiple call signatures, this compiles fine to JS

Basically my idea is equivalent to this, except you can call DiceRollerRoll statically at compile-time with the full type benefits of having unique names, and at runtime there is no call signature for DiceRollerRoll, and DiceRollerRoll would be created implicitly by the compiler (perhaps extending from some kind of ``Overload`` type)

Really it *is* the same as just creating multiple methods with different names as a suffix, just without the ugliness that is almost certain to be produced from keeping them unique from everything else, and slightly less verbosity

```dart
abstract class DiceRoller {
  DiceRollerRoll get roll;
}

abstract class DiceRollerRoll {
  final DiceRoller roller;
  DiceRollerRoll(this.roller);
  List<int> parse(String expression);
  List<int> many(int amount, int sides);
  List<int> from(DiceRoll roll);
}
```

One issue i could see with this syntax is being ambigous with a constructor where the overloaded method shares the same name as the actual type, but just like set/map literals it could prioritize the constructor where valid (i.e. initializing formal parameters, an initializer list, or redirecting to a constructor, aswell as where the body would initialize all non-nullable non-late fields) otherwise try an overload with dynamic return, or maybe overloads would need a return type to be recognized (that way, you don't conflict with constructors, and as a bonus you don't have an implicit dynamic return type), or simply disallow overloading the same name as the class to avoid this problem altogether?

# Operators

Personally i don't actually think method overloading would add that much to the language (though it would be nice), but i would really want to overload multiple signatures of the same *operator*, if you look at the implementation of + for ``int``, you'll find it inherits from ``num``, thus ``num operator +(num other)`` is the signature - however, ``int + int`` evaluates to an ``int``, even though that's not what the operator says it does, almost as if int has 2 signatures for +: ``int operator +(int other)`` and ``num operator +(num other)``. This makes sense, because if you add 2 integers, you won't get a fractional value, thus it's safe to say it's an integer. This isn't possible to implement in pure dart, this is simply a part of the type system and int being one of the core types, so therefore it gets special treatment, just like num is really a union of ``int|double``, FutureOr<T> is literally just a union type, and ``bool`` can't be extended or instantiated, it's just the type shared by 2 objects, all 3 of these examples make perfect sense, but are also impossible to implement in pure dart in the same way.

I'd imagine that for operators, there might not be as much overhead for multiple implementations based off the signature, since they're always restricted to a single required parameter, and overload resolution could essentially be simplified to ``dynamic operator +(dynamic other) => implementations[other.runtimeType](other);`` where ``implementations`` is a map, of course then i guess it's based on the runtime type and not the static type, so if an implementation used ``num`` it couldn't take an ``int``, so obviously something a little more sophisticated, but certainly a hell of a lot faster than needing to look up *multiple* parameters with signatures, aswell as optional parameters which more or less conflict with overloading, cause if you have ``range(int min, [int max = 100])`` and ``range(int max)``, which one is called with ``range(3)``?

I'd imagine that statically, it should look for the closest type in the supertypes of the operand's type (i.e. first checking if anything takes ``int``, then ``int?``, then ``num``, then ``num?``, then ``Object``, then ``Object?``, and finally ``dynamic``, and if the operand is ``Never`` then no implementation needs to be chosen at all), and dynamically (for faster resolution), it could look for in some given order the first implementation that supports the runtime type of the object (perhaps simply the order it's written in the class with subtypes being before their supertypes, and something more like ``operator +(other) => implementations.firstWhere((impl) => other is impl.operandType, orElse: () => throw TypeError());``)

With multiple operator overloads per operator, equality would be less of a pain to implement

```dart
class Foo {
  bool operator ==(Foo other) => true;
}

void main() {
  assert(Foo() == Foo());
  assert(Foo() != Object()); // Inherited from Object, and these are not the exact same instance
}
```

and i'd also love to be able to add extension operators that are already implemented in the class, for other types, but that isn't possible right now

```dart
extension BigIntArithmetic on int {
  BigInt operator +(BigInt other) => BigInt.from(this) + other;
}
extension IntArithmetic on BigInt {
  BigInt operator +(int other) => this + BigInt.from(other);
}

void main() {
  print(BigInt.one + 1); // 2
  var x = BigInt.one;
  x++; // Incrementing on BigInts! heck yeah!
  print(x); // 2
}
```

Maybe a restriction for this would be that you can't overload for a subtype of what's already overloaded in extension methods? but a really great use case would be defining custom equality rules for a class that has never heard of your library via extension methods, and that would break this use case, albeit only for the == operator specifically - maybe an alternative could be allowing extension methods to add another implementation for an operator if, and only if, the only operator in the superclass that supports the type is of ``Object``, and the given type will always return false in that implementation, a common pattern in equality operators is to check for each type before doing anything, otherwise returning false, so in a chain of such type checks, if the extension's operand type skips all of them and will be false, the overload is accepted for an extension

```dart
class Foo {
  final num value;
  Foo(this.value);
  bool operator ==(Object other) {
    if (other is num) return value == other;
    return false;
  }
}

extension FooExtensions on Foo {
  // Allowed: the catch-all implementation can only ever return true if it has a num, which a BigInt isn't
  bool operator ==(BigInt other) {
    return BigInt.from(value) == other;
  }
  
  // Error: int is a num, and a num could potentially return true in the catch-all, so this is *not* allowed
  bool operator ==(int other) {
    return value == other;
  }
}

class SubFoo {
  // On the other hand, this is completely fine, because it's overridden in a subclass, and as such doesn't change the behaviour of Foo based on what extensions you import...
  bool operator ==(int other) => value == -other;
}
```

A big benefit i see is any numerical type could support the increment and decrement operators aswell as being able to add its own type with the + operator, and maybe just in case this is the only reason it needs to support int, there could be some annotation that adds a warning to any code using the + or - operators with an int operand, instead of with ++ or --

**Comment by tatumizer** [comment]
Not sure whether this idea was suggested before, but anyway: if each variant of the method has one unique **required named** parameter (or, generally, a unique combination of required named parameters), then we have a method of resolution which is based on names only and doesn't make use of types at all (thus avoiding all associated problems). Complexity-wise, this is no different from having uniquely named methods, but the parameters are easier to name (with nouns) than the method (with a mangled verb). Example:

```
abstract class DiceRoller {
  List<int> roll({required String expression});
  List<int> roll({required int amount, required int sides});
  List<int> roll({required DiceRoll roll});
} 
```

Tear-offs won't work though, we need a special syntax for them, but tear-offs don't work for other things (like constructors) either. Perhaps we can find a syntax for tear-offs that covers all outstanding cases - including overloaded methods?

**Comment by passsy** [comment]
This would really help for deprecations when arguments change

```dart
@Deprecated('Change the index and element')
Iterable<E> whereIndexed(bool Function(E element, int index) predicate){ }

// new signature with index and element swapped
Iterable<E> whereIndexed(bool Function(int index, E element) predicate){ }
```

**Comment by jodinathan** [comment]
Is this at sight?  

All this `nonNull` versions of stuff is really boring and makes the code sparse and ugly.  

When using graphql, which we use a lot, most of things can be null and that is the point of its flexibilization. But we could still have sound stuff. 

For example in our listing function you can fetch the `totalCount` of items and that would be a `count()` in the database, which is expensive. So it makes a lot of sense to be optional, thus, null.  

We already use builders extensively and it makes sense for us to create several optimized methods that do the same but with different approaches or results.  

With the advent of static meta programming the source generated stuff shall increase even further.

**Comment by leafpetersen** [comment]
> Is this at sight?

This is not currently on our short to medium term road map.

**Comment by lukepighetti** [comment]
> Currently, a Dart object has at most one accessible member with any given name. Because of that, you can do a tear-off of a method. If you could overload methods, tear-offs would no longer work. You would have to say which function you tore off, or create some combined function which accepts a number of different and incompatible parameter signatures. It would make dynamic invocations harder to handle. Should they determine that method to call dynamically? That might cause a significant code overhead on ahead-of-time compiled programs.

Is there no clear way to have the compiler use it's knowledge of types to route a tearoff to the correct method based on the types associated with the field and the methods?

```dart
abstract class Foo {
  static void hello();
  static String hello();
}

main(){
  final String Function() secondHelloAsTearoff = Foo.hello;
}
```

**Comment by lrhn** [comment]
@lukepighetti Answer: ... maybe!

Using the context type may allow a tear-off to be chosen from the *known* members of that name.
For example:
```dart
class C {
  int foo(num x) => x.toInt();
  int foo(int x, int y) => x;
}
void main() {
  int Function(int) f= C().foo;  // chooses unary `int foo(num)`.
}
```
The problems comes when you have partial information available.
```dart
class D extends C {
  int foo(int x) => x;
}
void main() {
  C c = D();
  int Function(int) f = c.foo;
}
```
First question is whether `int foo(int)` is a different function than `int foo(num)`?
It's not a valid override, so it must be!
But(!) if the two were declared in the opposite order, would they still be different functions or would it be a virtual override?
(Whoops, what does overloading even mean in a language with covariant overriding? And optional parameters! That's not clear at all.)

Second question is whether the tear-off `c.foo` would choose the more precise `foo` from `D`, or the `foo` it knows is there from `C`? That is, is tear-off statically decided or dynamically decided?

Likely statically decided. That's definitely the safest approach. Which means needing to know the static type you're tearing off from, so what about tear-offs from `dynamic`.
```dart
void main() { 
  dynamic c = D();
  int Function(int) f = c.foo;
}
```
Does that even work? 
It does now, but we could say it doesn't work if `foo` is overloaded at all. 
Or we could dynamically try to find the best version for the context type.

And what about extension methods. Should we choose a more precise extension method over an instance method? (We will just have to make a decision, but likely yes.)

All in all, I think context type based tear-off selection is *possible*. It might not always be *predictable*, but that's because we don't have a real model for what overloading means yet. We might need a C#-like `new`/`override` marker to distinguish a potential override from an overload:
```dart
class C {
  int foo(int x) => x;
  int bar(int x) => x;
}
class D extends C {
  override int foo(num x) => x.toInt();
  new int bar(num x) => x.toInt();
}
void main() {
  C c = D();
  int Function(int) f1 = c.foo;  // D.foo by virtual lookup
  int Function(int) f2 = c.bar; // C.bar because D.bar is not the same function.
}
```

Decisions!

```

**Comment by munificent** [comment]
> All in all, I think context type based tear-off selection is _possible_. It might not always be _predictable_, but that's because we don't have a real model for what overloading means yet.

It's definitely possible because C# supports both overloads and tear-offs. Like you suggest, it goes off the context type. When you tear-off a method, it must be a context that unambiguously selects which overload you mean. Given that C# also supports type inference and user-defined conversion methods, that process is notoriously hairy... but it works.

**Comment by lrhn** [comment]
Yes, C# is the reason I think it can possibly work, but the reason their overloading works is an intricate set of decisions throughout the C# design. Tear-offs work too, but that's because they can already do type based resolution of overloaded methods.
If we can get overloading working, I think we can get tear-offs to work too. Getting overloading working is non-trivial.

Where Dart differs in a significant way is our optional parameters, which is really the Dart approach to overloading: One function declaration with multiple (related) signatures.

C# allows parameters to have default values, but that makes the *argument* optional, and the default value is inserted statically at the call point. A function like `void F1(int a, int b = 0)` is a binary function which you can call without providing the second argument. If you tear it off as a delegate like `public delegate void Binary(int a, int b = 7);`, then calling that delegate with one argument makes the second argument be 7. 
You cannot assign `F1` to a `public delegate void Unary(int a);` delegate type, it's inherently a binary function, it must be called with two arguments (one of which may be supplied by the compiler).

Dart's function subtyping, where we allow a `void Function(int, [int])` to *be* a `void Function(int)`, might get in the way of overloading.

**Comment by nex3** [comment]
Given that optional arguments act like overloads under some circumstances, one way to make the two play nicely would be to be fairly strict about their interactions—for example, I think it would be easy for a user to understand that `void Function(int)` and `void Function(int, [int])` can't both be defined for the same class.

**Comment by Levi-Lesches** [comment]
Combining your two comments, how about a rule that you cannot overload a member `a` with a subtype `b`, because in every instance where you can use `b`, you can already use `a`, so the compiler wouldn't know which one to use.

**Comment by Ryacinse** [comment]
> 

But Kotlin has both optional parameters and overloading. And optional parameters in Kotlin is more flexible.

**Comment by eernstg** [comment]
> And optional parameters in Kotlin is more flexible

The situation is somewhat mixed.

For instance, Kotlin function types are typed statically as `FunctionN<R, T1 .. TN>` (for a function accepting N parameters), and there is no subtype relationship between functions with a different number of formal parameters. This means that Kotlin can't safely abstract over functions which are callable using the same argument list shape (e.g., functions that accept one positional parameter of some type `T`), but where the run-time type is actually a function whose type accepts some optional parameters. In general, a Dart function type can be a subtype of another Dart function type because it accepts a larger set of optional parameters, but Kotlin can't support that.

This also prevents overriding method declarations in Kotlin from adding new optional parameters (you'll get the error that it doesn't override anything).

So there are several ways in which Dart is more flexible than Kotlin. Kotlin can support a completely untyped version of this kind of abstraction using reflection (so you can obtain the arity of a `Function` and then call it with that many parameters, and you won't get any default values), but Dart does it smoothly and in a way which is statically type safe.

**Comment by subzero911** [comment]
+1 for function overloading
You've added rather complicated language concepts like strong typing, covariants, enum classes, null-safety, and you even heading to add ADT.
But you didn't add such a basic OOP feature like function overloading because "it will make Dart difficult", that's just ridiculous!

**Comment by lukepighetti** [comment]
My preference, and it may be oversimplifying things, is to add type and parameter names to method signature


```
class Foo {
  bool getDayOfWeek({required DateTime fromDateTime}) {}

  /// collides with previous method, because the signature does not extend into the params
  /// would be nice if this was treated as a separate method from the one above
  bool getDayOfWeek({required Date fromDate}) {}
}

```

**Comment by Levi-Lesches** [comment]
@lukepighetti, types are already considered part of the signature. It's how Dart lets you know when you make an invalid override. But throwing overloading into the mix can make things complicated. Consider the following classes: 
```dart
/// A pair of 2D coordinates that can be decimals or integers. 
class Coordinates {
  final num x, y;
  const Coordinates(this.x, this.y);
  
  // Adds two coordinates together.
  operator +(covariant Coordinates other) => 
    Coordinates(x + other.x, y + other.y);
}

/// A pair of 2D coordinates that can only be integers. 
class IntegerCoordinates extends Coordinates {
  @override
  final int x, y;
  const IntegerCoordinates(this.x, this.y) : super(x, y);
  
  /// Only adds IntegerCoordinates together. 
  ///
  /// This ensures the resulting coordinates are also integers. 
  @override
  operator +(IntegerCoordinates other) => 
    IntegerCoordinates(x + other.x, y + other.y);
}
```

This is a common pattern where you subclass a type to make it more restricted. In this case, `IntegerCoordinates` can only contain integers and can only be added to other integers to keep the condition met. In today's Dart, that gives you this: 
```dart
void main() {
  final a = Coordinates(0.5, 0.5), b = Coordinates(1.5, 1.5);
  final c = IntegerCoordinates(1, 1), d = IntegerCoordinates(2, 2);
  
  // [Coordinates] can be added to any other subtype of [Coordinates].
  print(a + b);  // Coordinates(2.0, 2.0)
  print(a + c);  // Coordinates(1.5, 1.5)
  
  // [IntegerCoordinates] can only be added to themselves. 
  print(c + d);  // IntegerCoordinates(3, 3)
  print(c + a);  // Error: Coordinates cannot be assigned to IntegerCoordinates
}
```
With overloading, does this still hold? Or is `IntegerCoordinates.+` no longer an _override_ of `Coordinates.+` but rather a whole new method, so that `c + a` is now valid?

----

Dart isn't Java, but since Java has overloading and overriding, I thought I'd bring a Java example to compare. Due to Java using a different type of number system than Dart, I repurposed the example into one that makes less logical sense but still shows the problem: 
```java
class Car {
  void crash(Car other) { System.err.println("Two cars crashed!"); }
} 

class Boat extends Car {
  // Turns out this is an *overload*, not an override, of Car.crash. 
  void crash(Boat other) { System.err.println("Two boats crashed!"); }
}

public class Temp {
  public static void main(String[] args) {
    final Car car1 = new Car(), car2 = new Car();
    final Boat boat1 = new Boat(), boat2 = new Boat();

    car1.crash(car2);  // Car.crash
    car1.crash(boat2);  // Car.crash

    boat1.crash(boat2);  // Boat.crash
    boat1.crash(car1);  // This shouldn't work, but it calls Car.crash 
  }
}
```

**Comment by Jetz72** [comment]
> What would happen if we removed the unique name requirement and did nothing else?

For one, without specifying any rules to decide how to prioritize or disambiguate them, determining what gets called becomes unclear:

```dart
void foo(int x, num y) => print("foo type 1!");

void foo(num x, int y) => print("foo type 2!");

void main() {
  foo(1, 2.0); //All is well, calls type 1.
  foo(3, 4); //Which one do we call here?
}
```

Tear-offs are a bigger issue:
```dart
var x = foo; //Which one? They both could fit into an int-int function type but what exact type is x?
if(x is void Function(int, num)) print("It's type 1!");
else if(x is void Function(num, int)) print("It's type 2!");
else if(x is void Function(int, int))
  print("It's not a reference to the function itself, it's some placeholder that kicks the discrepancy down the road "
  "to the invocation site and has an inexact type until then!");
else print("It's still a placeholder, but either we aren't computing the bound of foo's overloads or someone added "
  "another overload for foo which subtly changed the behavior here simply through its existence!");
```

**Comment by munificent** [comment]
Overloads are a useful, powerful feature. But they are also an enormous can of worms that add a ton of complexity to the language in ways that can be confusing and painful for users. Some examples off the top of my head:

### Tear-offs

Presumably this is fine:

```dart
class C {
  bool method(int i) => true;
}

main() {
  var tearOff = C().method;
  var x = tearOff(3);
}
```

Now say you add an overload:

```dart
class C {
  bool method(int i) => true;
  String method(String s) => s;
}

main() {
  var tearOff = C().method; // ?
  var x = tearOff(3); // ?
}
```

What is the type of that `tearOff` now? Is the call to it valid? If so, what is the type of `x`?

We *could* say that it's an error to tear off a method that has overloads. But one of the primary *goals* with overloading is that it lets you add new methods to a class *without* breaking existing code. If adding an overload causes untyped tear-offs to become an error, then adding any overload anywhere is a potentially breaking change.

We could say that it's *always* an error to tear-off a method in a context where there is no expected type. That would be a significant breaking change to the language. That also means that changing an API to loosen an expected type (for example changing a callback parameter from a specific function type to `Function`) is now a breaking change.

### Overriding and parameter types

Consider:

```dart
class A {
  void foo(int i) {}

  void bar(String s) {}
}

class B {
  void foo(num i) {}

  void bar(bool b) {}
}
```

Is `B.foo()` an override? Presumably yes. Is `B.bar()`? Probably not?

So what if you later change B to:

```dart
class B {
  void foo(num i) {}

  void bar(Object o) {}
}
```

Now that presumably will become an override of `A.bar()`. So the parameter types of a method may affect whether or not its overriding a base class method.

### Optional parameters

Consider:

```dart
class C {
  void method(int x) {}
  void method([int x, int y]) {}
}

main() {
  C().method(1);
}
```

Is that a valid set of overloads? If so, which one is called when you pass a single argument?

### Promotion

Consider:

```dart
class C {
  void method(num n) { print('num'); }
  void method(int i) { print('int'); }
}

test(num n) {
  if (n is int) C().method(n);
}
```

I'm guessing this prints `int`. Are there other interactions between overload selection and type promotion that are less intuitive? Do users need a way to opt *out* of having promotion affect overload resolution?

### Inference

Consider:

```dart
class C {
  void method(List<int> ints) {}
  void method(List<Object> objects) {}
}

main() {
  C().method([]);
}
```

What does this do?

### Generics

Consider:

```dart
class C<T> {
  method(num n) {}
  method(T t) {}
}
```

Is that a valid pair of overrides? What if `T` is instantiated with `num`? `int`? `Object`?

What about:

```dart
class C<T extends num> {
  method(String s) {}
  method(T t) {}
}
```

Are those now valid overloads?

What about:

class A<T extends num> {
  method(T t) {}  
}

class B<T> extends A<T> {
  method(num n) {}
}
```

Is `B.method()` considered an overload of `A.method()` or an override?

I'm sure it gets even weirder when generic functions come into play.

### Covariant

Here's a fun one:

```dart
class A {
  void method(num n) {}
}

class B extends A {
  void method(covariant int i) {}
  void method(Object? o) {}
}
```

Is the second `method()` in `B` a distinct overload, or does it collide with the other `B.method()`?


The point is not that all of these issues are unsolvable, but it's that they *must* be solved in order to add overloading to the language and users will have to *understand* at least a fraction of them whenever they interact with the feature.

**Comment by jodinathan** [comment]
@munificent two things:  

1) Dart already has overloading in the `try catch` clause. If overloading is not something necessary, don't you agree that all `catch` should be something like `catch(Exception e, [StackTrace st])`?  

2) Is there any plan to not have these many `firstWhereOrNull` without losing performance? I mean that with overloading the compiler do a check and call the correct method by the type signature so we don't have to do it at runtime.  
Also a simple rule that we have here is not to have methods with `And` or `Or` in the name. It usually means bad architecture and the whole context should be reanalyzed.

**Comment by jodinathan** [comment]
> which I'd do in a heartbeat if I could migrate the world easily

The point is not if we can change that or not, but the reasoning behind not having an optional parameter in the first place.  

Assuming that not requiring the `StackTrace` in the `try catch` clause can be seen as an optimization, since we are telling the compiler we don't need that info, makes total sense to it be overloadable as it currently is.

The point is simplicity. You don't want to see the `StackTrace` argument everywhere as you don't want to have `firstWhereOrNull` kind of method.

Be it parameter or return type, the point is overloading in general. However, if we could have at least the return type overloadable I would be glad, honestly.

**Comment by Levi-Lesches** [comment]
> 1\. Dart already has overloading in the `try catch` clause. If overloading is not something necessary, don't you agree that all `catch` should be something like `catch(Exception e, [StackTrace st])`?

@munificent brought up valid questions and complications that arise with methods, so comparing it to try/catch isn't a valid response because it doesn't answer those questions. 

> 2\. Is there any plan to not have these many `somethingOrNull` like `firstWhereOrNull` without losing performance?

I agree with this but the ideal answer, IMO, would be to make `firstWhere` nullable in the first place to cover all cases, not overloading.

> The point is simplicity. You don't want to see the `StackTrace` argument everywhere as you don't want to have `firstWhereOrNull` kind of method.

Simplicity is good, but the sheer amount of syntactic and semantic questions that arise when using overloads negates this. If you have to ask yourself which method you're using, I don't see how that's simpler than `firstWhere`/`firstWhereOrNull`. Sure, the latter is more verbose, but you never have to ask yourself what arguments/return values are valid because it's in the name. Doesn't get much simpler than that. Overloads, however, mean you always have to check the source code to see which types you can use because there may be more than one implementation with the same name. 

For example, imagine parsing a `DateTime`. Right now there are several options: `.fromMicrosecondsSinceEpoch`, `.fromMillisecondsSinceEpoch`, `.utc`, `.parse`, and `.tryParse`. Sure these names are quite verbose, but imagine trying to figure out what value to pass if there were just one `DateTime.parse` constructor.

**Comment by jodinathan** [comment]
> For example, imagine parsing a DateTime. Right now there are several options: .fromMicrosecondsSinceEpoch, .fromMillisecondsSinceEpoch, .utc, .parse, and .tryParse. Sure these names are quite verbose, but imagine trying to figure out what value to pass if there were just one DateTime.parse constructor.

The only one of that list that I see using overload is `parse`.  

The reason why we have `firstWhereOrNull` instead of `tryFirstWhere` is because when you type `firstWhere` you see the `firstWhereOrNull` in the list. It would be harder to find the nullable version if it was `tryFirstWhere` because of autocomplete.  

I understand the point regarding the aesthetic preference but think with me a bit and take the `tryParse` above as example.  
Coding something we know that there is a `something` method, now we need its nonnull version. What is easier?
  
a) a method that is overloaded to nullable and nonullable versions that you can read the docs and the implementation in the same place
b) some random name that is up to the developer like `trySomething`, `somethingOrNull` or `nonNullSomething`

**Comment by TheKashe** [comment]
> Tear-Offs

How about

```
typedef bool Foo(int i);

class C {
  bool method(int i) => true;
  String method(String s) => s;
}

main() {
  Foo tearOff = C().method; //will resolve to: bool method(int i)
  var x = tearOff(3); // true
}

//or perhaps
main() {
  var tearOff = C().method<Foo>; //will resolve to: bool method(int i)
  var x = tearOff(3); // true
}

//or without the typedef
main() {
  var tearOff = C().method<int>; //will resolve to bool method(int i)
  var x = tearOff(3); // true
}

```

**Comment by lrhn** [comment]
Using the context type to disambiguate the tear-off operation has been suggested.

It differs from the other places where we currently use the context type to do something. In those, the context type can cause something to happen (`call` tear-off, generic instantiation, downcast from `dynamic`, etc.), otherwise there is syntax to make the same thing happen explicitly instead, if you have no context type, or don't want to rely on it.
For overloaded method tear-off defined like this, there is no similar thing. There is no explicit tear-off syntax.
(You suggest `C().method<Foo>` here, but that's already valid syntax for explicitly instantiating a generic method tear-off.)

If we make `as` introduce a context type (which we should!), then `C().method as Foo` would be an explicit way to choose the context type, and therefore the method to tear off.

(There are other issues with method overloading in Dart than tear-off resolution. For example, Dart's optional parameters is already a kind of overloading, which allows a single method to satisfy two different member signatures. When you combine those with actual overloading, and subclassing, it's non-trivial what it even means to override one or more superclass members.)

**Comment by mattrberry** [comment]
On the topic of using a context type to disambiguate, there's also Crystal which requires you to specify the types of the parameters in order to tear off a method

```crystal
str = "hello"
proc = ->str.count(Char)
proc.call('e') # => 1
proc.call('l') # => 2
```

https://crystal-lang.org/reference/1.6/syntax_and_semantics/literals/proc.html#from-methods

**Comment by dcharkes** [comment]
> Dart's optional parameters is already a kind of overloading

But in practise we advice making more members instead of using this kind of overloading, especially if we can't statically enforce the right arguments

```dart
void foo({required A a, B? b, C? c}) {
  if(b != null && c != null) throw ArgumentError('Provide at most one of b and c.');
  // ...
}
```

We usually rewrite that to two methods (which have overlapping behavior if neither b nor c are provided).

Now, if we could write some kind of more fancy constraints of permutations of optional parameters that are correct that would work with static checking ... 🎉

**Comment by Jetz72** [comment]
An approach occurred to me that might at least handle the overload/override ambiguities, tear offs, and mitigate the possibility of unintentional breaking changes - named overloads:
```dart
class A {
   //We name each overload at the declaration site.
   String foo.withInt(int x) => "int (A)";
   String foo.withString(String x) => "String (A)";
   
   String foo() => "Nothing"; //Not allowed, name collision on "foo" and can't be an overload of it without a name.
}

class A1 extends A {
   //To override one, you must use the same name.
   @override
   String foo.withString(String x) => "String (A1)";
   //To make a new overload in a subclass, just use a new name.
   String foo.withNum(num x) => "num (A1)";
}

void main() {
   var a = A1();
   //The name is not required at the call site.
   print(a.foo(1)); //int
   print(a.foo("hello")); //String (A1)
   print(a.foo(1.0)); //num (A1)
   //A specific overload can be referenced by name, which is used to perform tear-offs safely, or force a particular overload.
   var bar = a.foo.withInt;
   print(bar(1)); //int
   print(a.foo.withNum(2)); //num (A1)
}
```
Requiring each overload to have a unique label helps with declaration ambiguities while letting the variants share a singular name at the call site. I do realize the silliness in solving a chunk of the problems with overloading by giving the overloads distinct names - i.e. missing the whole point of overloading. But in practice it seems like only a slight sacrifice in convenience assuming you write a method once then use it many times going forward.

This doesn't address all the problems, of course. Dispatch in particular would still have some pretty glaring questions, and I'm not really convinced that this feature would be worth the trouble in the first place. But I wanted to jot down this line of thought and see if it goes anywhere when fleshed out.

Going down the above (not comprehensive) [list](https://github.com/dart-lang/language/issues/1122#issuecomment-1169237638) of things that would need to be solved, it at least provides solutions for these dilemmas:

> ### Tear-offs

Since tear offs have to be manually disambiguated, adding a new overload with its own label won't affect existing tear offs. It would be a breaking change to make a non-overloaded method into an overloaded one, but the change in declaration syntax should make that obvious enough. 

You could maybe add a way to blindly tear off an overloaded function and then resolve it at the call site - either storing it dynamically and letting the user assert that it'll work at runtime, or keeping it in a special function type that remembers the original overload family as part of its static type - but that seems tricky either way.

> ### Overriding and parameter types

```dart
class A {
  void foo.a(int i) {}
  void bar.a(String s) {}
}

class B {
  void foo.a(num i) {} //Valid override
  void bar.a(bool b) {} //Invalid override, since String isn't a subtype of bool.
  void bar.b(bool b) {} //Valid new overload, not an override.
  void bar.a(Object b) {} //Valid override
}
```

> ### Covariant
```dart
class A {
  void method.a(num n) {}
}

class B extends A {
  void method.a(covariant int i) {} //Valid override
  void method.b(Object? o) {} //Valid new overload
}
```
That doesn't answer the question of which method a double would dispatch to, especially if the static type is A but the runtime type is B. Similar complications remain with generic parameters and dynamic invocation, and I've barely given any thought to how return types work with all this, so I'm inclined to just leave it at that for now.

**Comment by lrhn** [comment]
The extra name sounds like multiple different-named methods which allow being called through a (shared) prefix of the name, using some sort of static resolution. 
It has to be static resolution because the resolution affects type analysis when the return types can differ, and runtime then has to match the statically predicted behavior to ensure soundness.

There is no real overloading, there are just longer names and static dispatch to one of those longer names when you only specify a prefix of the longer name.

You can add more members with the same prefix in a subclass. Static resolution won't see them when calling on the superclass. And you can always call a specific method directly.

Overloading an existing member gets weird. If you have a `foo` method already, adding a second `foo.second` method gets hard to specify.
We could say that `foo` is short for `foo.<empty>`, and it always exists in parallel with the other members with the same prefix, but then you either lose the ability to call that member explicitly, or  to call `foo.second` as `foo`, because `o.foo()` can only mean one of those. We could introduce a special token, `o.foo.#empty()` to signify the empty name, like we have `.new` on constructors. Maybe just `o.foo._()`. (Or would that be a library-private sub-name? It probably should be, so you can have `foo.public()=>...; foo._private()=>...` and only expose one publicly.)

In any case, it's like `o.foo` is a callable object (like an object with a `call` method), but it has *multiple* ways to be called, one for each `foo` suffix member declared. 

The thing you get for free is that invoking that object as a function will statically decide which of the members is called.

Tear-offs can be decided the same way as invocations, using a context type's parameter types to choose instead of actual argument types. If there is no useful context type, it's an error.

This has all the usual problems with `dynamic`. You just have to use the full name there.

It also suffers from the same kind of name conflicts as Dart does today. If you add `int foo.x(int i) => i;` in a subclass, and the superclass later adds `String foo.x(String x) => x;`, you have a naming conflict.
Full type-based overloading would consider those different members because they have different types.

**Comment by Jetz72** [comment]
> Overloading an existing member gets weird. If you have a `foo` method already, adding a second `foo.second` method gets hard to specify.

The way I was imagining it, it'd be an error to have an overload syntax `foo.second` alongside a plain `foo`; converting a normal function into an overloaded one would be a breaking change from adding the label, but adding further overloads beyond that would be fine (barring that usual name collision with pre-existing subclasses thing).


> It has to be static resolution because the resolution affects type analysis when the return types can differ, and runtime then has to match the statically predicted behavior to ensure soundness.

Ah, I figured once return types were considered the whole thing would snag pretty fast. I was sorta leaning towards dynamic invocation where you resolve all the runtime types before picking a method. I had a hard time imagining how you'd be able to statically handle the covariant situation, generics, and dynamic parameters. But I guess that would just propagate the problem outwards. Even if you computed a least upper bound for all the possible return types (and good luck when subclasses can add new ones), it'd still make it a strongly breaking change to introduce new overloads, which was something I was hoping to avoid. Losing static analysis for the whole feature would pretty strongly tip the "is it worth the trouble" scales towards the "no" side.

**Comment by jodinathan** [comment]
While I was creating [Typings](http://pub.dev/packages/typings) I had to deal with transpilling overloaded methods from TS to Dart. As JS interop uses extension methods a lot, I thought of a solution for tearoff and also bad overloading usage:  

### Allow overloading methods only in extension methods.

So a class:

```dart
class Foo {
  void bar();
}
```

Could have `bar` overloaded:

```dart
extension OverFoo on Foo {
  void bar(String daz); // ok
}
```

Then you can easily catch the `bar` version you need:

```dart
final foo = Foo();
final v1 = foo.bar; // void bar() {}
final v2 = (foo as OverFoo).bar; // void bar(String);
```

With this approach you can statically know the correct callback while also putting a minor restriction to its usage without really restricting it.  

It would also solve one the [JS interop questions](https://github.com/dart-lang/language/issues/2995)

**Comment by lrhn** [comment]
Allowing overloading only using extension methods was actually something I intentionally tried to prevent with the extension methods design. It was a goal (not completely reached, but still an aspiration) to not make extension methods *better* than instance methods.
That would cause users to design their APIs around extension methods instead of instance methods, with the short-comings that also has (no virtual dispatch being one).
Extension methods are available as an alternative to instance methods for some uses-cases, but not as a patch for reasonable functionality that instance methods could have too.

If we want to allow overloading, we want to allow overloading on both instance methods and extension methods, so the choice between instance and extension method is made on the basis of which best fits the use case, not which of them happens to allow overloading.

Anyway, if we were to allow overloading on extension members, then we'd still infer the type of the receiver, then check which extensions apply to that, and whether they have one (or now more) members of that name.

Then, most likely, instead of just picking the best `on` type from the applicable extensions, we'd also try to choose the best method based on the *arguments* (or even return type).
That's where it gets tricky, because we can't properly do type analysis on the argument expressions without knowing the context type provided by parameter types of the function being called, but we don't yet know which function we are calling.
So the best we can do to disambiguate is to eliminate all extension methods which definitely *cannot* be called with the provided number of positional arguments and names of named arguments. That's not a particularly good overloading resolution approach, it can distinguish between `String pathFrom(String)` and `String pathFrom(Uri)`, which is like table-stakes for having overloading.

But nothing about that is special to extension methods, it applies equally well (and badly) to instance methods,
and that's one of the reasons it's so darn hard to add overloading to Dart. The type inference algorithm assumes that the method being called is known *before* type analyzing the argument expressions, and therefore the types of arguments cannot be used to choose which function to call.

Maybe that choice in type inference was a bad idea, but it's not going to be easy to change.

In short: Nothing here is unique to extension methods. Anything that can be used to resolve overloading involving extension methods can also be used to resolve overloading involving instance methods as well. And it has all the same problems which make that particularly hard for Dart.

(Also, `(foo as OverFoo).bar` is already possible as `OverFoo(foo).bar`.)

**Comment by jodinathan** [comment]
I thought the motives for not having overloading would be 
- By design: If spammed can make ugly/hard to read code (C# comes to mind)
- Technical: Troublesome to tearOff

That is why I thought overloading through extensions would be a help there. 

> The type inference algorithm assumes that the method being called is known before type analyzing the argument expressions, and therefore the types of arguments cannot be used to choose which function to call.

This would not be an easy change, however, doesn't the new inference system already crawl the method beforehand?  

Consider:  

```dart
enum SomeEnum<T> {
  a<int>();

  const SomeEnum();
}

K foo<K>(SomeEnum<K> bar);
```

Then when you use `foo` you get the correct infered type:  

```dart
final k = foo(SomeEnum.a); // k is an int
```

**Comment by lrhn** [comment]
Type inference of `final k = foo(SomeEnum.a);` works by:
* Infer the type of `foo(SomeEnum.a)` with no context type (because `final k` doesn't have any type expectations).
* Type analyze the receiver `foo`. It resolves to something with type `K Function<K>(SomeEnum<K>)`.
* Try to infer `K` from the context type. Since there is no context type, that fails.
* Type analyze `SomeEnum.a` with context type `SomeEnum<K>`.
* This resolves `SomeEnum.a` to its declaration which has static type `SomeEnum<int>`.
* If there had been further arguments, we'd have looked at those too (actually, now, incrementally in some order that  improves our chance of solving for multiple variables).
* Then try to solve for `K` again. Succes `K = int` found.
* Then go back to to the invocation `foo(SomeEnum.a)` and check that `SomeEnum<int>` is assignable to the parameter type `SomeEnum<K>` with `K = int`, which it is. So make the type parameter in the call be `int`, effectively inferring that the expression is really `foo<int>(SomeEnum.a);`.
* Then the static type of `foo<int>(SomeEnum.a)` is found to be `int`.
* Type inference infers the declared type of `k` to be `int`.

So, type analysis finds the static type of expressions. Sometimes those types include yet-unbound type variables.
At suitable points in the analysis, type inference tries to infer type arguments for those type variables. Then it continues.

That's all well and fine, but it relied on having the full function type of `foo` in step 2.

Now imagine there being multiple possible functions that `foo` can refer to:
```dart
int foo(SomeEnum<int> a) => 9;
T foo<T>(SomeEnum<T> a) => a.value;
String foo(Object? a) => a.toStrung();
```
For each one of these, type analysis and inference can proceed, and they're *all valid*. But the return type differs.

Then make it even harder, by not having an argument type which doesn't require any inference.
```dart
Final<int> foo(Future<int> f) => f;
Final<double> foo(Future<double> f) => f;
/// and
final k = foo(Future.value(0));
```
Now it gets ambiguous.
The type inference and static type of `Future.value(0)` depends on the context type. The static type of `0` depends on the context type.

With only `Final<int> foo(Future<int> f) => f;` in scope, `foo(Future.value(0))` would give `Future.value(0)` a context type of `Future<int>`, which will give `0` a context type of `int`, and type inference will turn it into:
```dart
final Future<int> k = foo(Future<int>.value(0));
```

With only `Final<double> foo(Future<double> f) => f;` in scope, type inference would make this effectively:
```dart
final Future<double> k = foo(Future<double>.value(0.0));
```
(Notice the meaning of the decimal literal depends on context type too!)

With both in scope, which one should be chosen? Either works, neither is more or less specific than the other,
and looking at the *argument expressions*'s static type is impossible, because the argument type depends on the function type that it's an argument of.

**Comment by jodinathan** [comment]
> Final<int> foo(Future<int> f) => f;
Final<double> foo(Future<double> f) => f;

We could be restrict those with an analyzer error `This overload leads to ambiguous methods signatures`

**Comment by lrhn** [comment]
> We could be restrict those with an analyzer error 

We could, but what would the criteria be for deciding whether two declarations introduce ambiguity or not?

I'm pretty sure I can always create an argument expression which depends on the context type, and do so in a way that makes it necessary to know the parameter type before doing type analysis on the argument expression.

The way inference works, just a
```dart
T hack<T>(List<T> list) => list.first;
// ...
  anyFunction(hack([null]));
```
will use the context type `C` to infer `hack<C>(<C>[null])` which is a compile-time error if `C` is non-nullable,
but valid if `C` is nullable.
If the compiler doesn't know the context type, it can't decide whether that can be compiled.
And it can't decide what the static type of `hack([null])` is before it knows the context type, so it can't use that type to decide which function to call.

So, the only safe criterion for not causing a potential conflict is that the functions have completely incompatible argument *shapes*. And then we're back to not having type-based overloading, only shape-based.
Which is a valid choice, but probably just not what people really want.

**Comment by jodinathan** [comment]
> And then we're back to not having type-based overloading, only shape-based.
Which is a valid choice, but probably just not what people really want.

Would shape-based be number of arguments?  
If yes then it is really not what the users want.  

Quick thinking on how to try to fix the ambiguity problem...  

I would take two approaches, first would be at definition by flatting all types and check if they collide:  

```dart
void foo<T>(T bar);
void daz<K>(K yum); 

// flatted: 
void foo(dynamic bar);
void daz(dynamic yum);  // this throws an ambiguity error
```

Then at usage level we must make sure that there is only one definition from types analysis.
Definition: 

```dart
Final<int> foo(Future<int> f) => f;
Final<double> foo(Future<double> f) => f;
```

Usage: 

```dart
final k = foo(Future.value(0)); //  as the analyzer found two possible methods, 
// it throws an ambiguity error

// the user could possibly fix this by making sure the analyzer know all types: 
final k = foo(Future.value(0) as Future<int>);
```

**Comment by lrhn** [comment]
You get the same issue with just
```dart
void foo(int x) {}
void foo(double? x) {}
```
Those are completely disjoint parameter types, no generics, no shenanigans.

Again, then I just do:
```dart
T hack<T>(List<T> l) => l.first;

foo(hack([null]));
```
and you still can't tell me whether that program is correct or not, or which `foo` to choose.

To choose based the function based on the argument type, you need to know the static type of the argument expression.
To find that you need the context type (unless we stop providing context types for arguments, which will be a major blow to type inference) in order to analyze the expression in the context.
To get a context type you need to have already chosen a `foo` function.

So you cannot choose the the function to call based on the argument type.


We *could* choose to not provide the argument expression with a context type when there are multiple possible functions to call.
Or we could select the ones which have a viable parameter list shape, and find the least upper bound of the parameter types of all those functions, and use that as context type. (Remembering that least-upper-bound is a fickle and confusing operation that can easily end up giving you effectively no context type anyway, especially for overloading where the argument types are more likely to be unrelated than not.)

Then adding an overloading would worsen type inference for everybody. You'd be better off just giving it another name. (And then we're back to square one, which is probably why we've been here for a while.)

**Comment by tjarvstrand** [comment]
> So you cannot choose the the function to call based on the argument type.

IMHO it would be perfectly acceptable for the analyzer to emit an error when it doesn't have enough information to make a decision, as long as the error message is clear. This is what e.g. Scala does (and I imagine Kotlin too), and it works well.

**Comment by stan-at-work** [comment]
Why is this not implemented yet, its been 13 years ...

**Comment by esoros** [comment]
I'm pretty sure that languages that support var x = "foo" and anonymous methods (such as LINQ with C#) with overloads are going to push the compiler to algorithms that are in NP-Complete and then would just be too slow. Kotlin, Swift, Modern C#, Modern Java aren't known for quick builds. In the context of business with large orgcharts, managers will start assigning costs to builds I'd argue it's better if those are faster than slower.  As stated above, in the context of the call site, how would the compiler be able to quickly determine what methods to call (normally there's an entire express tree, not just one method?)

    If people are going to basically not have any kind of type inference (so, the developers type in all of the types, including the type of the output parameter), then it's probably not a big deal. Developers like to type in extremely literal types (especially with generics) so the staff may not want to type in ICustomDomainObjectLoader<DomainObject<CustomerOrderStartefy, CustomResolverStraget>>> repeatidly and then folks start getting into type inference algorithms. There's physical limits to such things and tradeoffs. I guess it just depends. If someone knows of a system that has quick inference times (i.e quick builds) and also allows methods with the same name and same number of parameters, please let me know. 

https://web.cs.ucla.edu/~palsberg/paper/dedicated-to-kozen12.pdf

**Comment by rubenferreira97** [comment]
I wanted to share a more distinct example to highlight that Dart lacks support for implementing the same generic interface multiple times with different type parameters:

```cs
using System;

interface IProcessor<T> {
    public string Process(T input);
}

class MultiProcessor : IProcessor<int>, IProcessor<double> {
    public string Process(int input) {
        return $"Processing integer: {input}";
    }

    public string Process(double input) {
        return $"Processing double: {input:F2}";
    }
}

class Program {
    static void Main(string[] args) {
        MultiProcessor processor = new MultiProcessor();

        // Explicit interface usage
        IProcessor<int> intProcessor = processor;
        IProcessor<double> doubleProcessor = processor;

        Console.WriteLine(intProcessor.Process(42));       // Output: Processing integer: 42
        Console.WriteLine(doubleProcessor.Process(3.14));  // Output: Processing double: 3.14
    }
}
```

Not sure how frequently this capability is used, but it’s a funny/notable feature that C# supports and Dart currently does not, especially without method/function overloading.

**Comment by TekExplorer** [comment]
I think that overloading is... Too useful in terms of API to be ignored - especially with augmentations.

Consider:
```dart
String foo(Ref ref) => ...;

// Generated
@overload
Provider<String> get foo;
// Or, if necessary
Provider<String> foo();
```
This allows for sharing the name, which has semantic bonuses to using `fooProvider`

Width overloading of this kind is pretty easy I would imagine. On its own, it's already a benefit, I think.

For stuff like
```dart
class Foo {
  void add(String? value);
  void add(List<int>? value);
}
```
`String` and `List<int>` are unambiguous, but that gets confusing if you pass null. Which does it pick?

In this case... Perhaps `add`'s type is a more useful `Function` that includes specific getters disambiguating each one?

That's sort of what was mentioned prior with `foo.withString` and `foo.withInt`, so perhaps defining names makes more sense.

Then again, maybe same-width "overloads" should just use union types

Extensions then, could be the only way to have an "actual" same-width overload that simply picks out based on type, and if it's ambiguous - let the compiler complain like normal.
```dart
class FooBad {
  bar(List value);
  bar(Map value); // error
}
class Foo {
  // Ok
  bar(List || Map value);
  bar(String value, int index);
}

class Foo2Bad extends Foo {
  bar(int value); // bad
}
class Foo2 {
  // Ok
  bar(List || Map || int value) {
    if (value is int) return;
    super.bar(value);
  }
}
extension S on Foo2 {
  // Ok. Normal disambiguation rules.
  bar(String value);
}

extension Foo2Int on Foo2 {
  // Ok, but prefers instance method
  bar(int value);
}

foo2.bar({});
foo2.bar([]);
foo2.bar(0);
foo2.bar('str'); // S
Foo2Int(foo2).bar('str'); // ok

extension L on Foo {
  add(List value);
}

extension M on Foo {
  add(Map value);
}

foo.add([]);
foo.add({});
// Ambiguity error
foo.add(ListMap());
// Ok
L(foo).add(ListMap());
```
What about tearoffs?

When the context function type is known, grab the one that works.

Always tear off the instance method if it exists first. As there is only one of each width, we could possibly disambiguate with a cast, or a synthetic getter that represents a particular type.

For example
```dart
final fn = foo2.bar.with<Function(List || Map || int)>;
```
`with`, in this case, accepts only a union of function types available on the instance.

It could also simply reuse `as` to "coerce" only the valid ones.

It returns `Never` if no valid fn is found.

How this interacts with extension types, that depends on if you prefer to merge same-widths (powerful, but also dangerous) or keep each one independent (likely safer)

This would be the easy-enough `Ext(obj).fn`

Extension types would have the same restrictions as objects in its own declarations - that being that you can overload with different widths, but same-widths need to be merged.

Alternatively, we allow non-merged, but let everyone else see it as merged, without a way to separate them - aside from explicit casts maybe.

Idk what do you think?

I'm sure I missed a bunch of stuff.

**Comment by rrousselGit** [comment]
Although overloads would be very useful for Riverpod, I don't think the examples given are right.

It wouldn't be about writing:

```dart
@riverpod
T function(Ref ref) => ...
```
Then generating:
```dart
final function = Provider(...);
```

Instead it'd be about changing the prototype of `Ref.watch` to:

```dart
class Ref {
  T watch<T>(T Function(Ref ref) provider);
  T watch<T, ArgT>(T Function(Ref ref, ArgT arg) family, ArgT arg);
  T watch<T>(Notifier<T> Function() notifier);
}
```

Used as:

```dart
T provider(Ref ref) => ..
T family(Ref ref, String arg) => ...
class MyNotifier extends Notifier<T> {...}

[...]
T value = ref.watch(function);
T value = ref.watch(family, 'my arg');
T value = ref.watch(MyNotifier.new);
```

The only way to do that today would be to have multiple `watch` methods. 

Like `watch` vs `watchFamily` vs `watchNotifier`.  
And then do the same for `ref.listen` / `ref.read` / `ref.exist` and anything else that takes a provider as parameter.


I don't think having so many variants of the same thing is good.

Instead Riverpod used code-gen to have its own form of function overloading.

**Comment by TekExplorer** [comment]
Right, fair enough.

Technically we can do this in reverse by having extensions on the relevant classes the way context_watch and friends do with `future.watch(context)` and `stream.watch(context)` etc

But that may not be a desired API, despite being the only option for it at the moment.

**Comment by rrousselGit** [comment]
Indeed. That's something I'm considering.
But I don't like how this looks with classes: `MyNotifier.new.watch(ref)`

That `.new` is ugly. And static extensions by themselves aren't enough to remove it.
We'd need generic static extensions, to write, combined with https://github.com/dart-lang/language/issues/4200

**Comment by rubenferreira97** [comment]
To add a bit of motivation for this feature, I recently watched a session from Google I/O 2025 on YouTube, and it immediately brought this issue to mind.

Since Dart doesn’t support overloading natively, JNIGen ends up generating method names like `doSomething$1`, `doSomething$2`, and so on, just to work around the limitation. While this works, it’s not great for developer experience (DX), as it makes the API harder to read, discover, and use.

Here’s a [quick demo](https://youtu.be/flwULzNYRac?si=ezpEaZoGX9840N5K&t=469) showing the issue in action.

Native support for overloading in Dart would make these bindings much cleaner and more idiomatic — not to mention other benefits discussed in this issue, such as enabling better API design.

**Comment by rrousselGit** [comment]
Coming back to this:

For quite some time now, the community has started using extension methods as a workaround to the lack of overloading.
It's fairly common to see a package do:

```dart
extension<T> on Stream<T> {
  void method() {}
}

extension<T> on Future<T> {
  void method() {}
}

extension<T> on ValueListenable<T> {
  void method() {}
}


stream.method();
future.method();
listenable.method();
```

Instead of:

```dart
method(stream);
method(future);
method(listenable);
```


This raises the question: **If static extensions already enable some pseudo-form of overloading, why not naturally support it?**

Static extensions are a bit flawed when used like this, as it leads to:
- poor error message when there's no matching overload.   
  Say we write `42.method()` with the extensions defined above, instead of a "No matching overload found", we get a "method not found".
- poor discoverability.    
  It's fairly hard to see the list of supported types with this syntax.
  In the previous example, we have Future/Stream/ValueListenable as supported, but there's no centralized place in the IDE to see them all.
  Autocompletion on `future.method()` will not tell you in any capacity that you can alternatively do `listenable.method()`
- Degraded readability.   
  Packages would often _prefer_ `method(obj)` over `obj.method()` in those situations, but they don't have the choice.
  Take `42.observable` vs `list.observable` as example, to convert an object into a `Observable` vs `ListObservable` using overloading.  It'd read better as `observable(42)` vs `observable(list)`.


In short, using extensions like this kind of has the opposite effect of what extensions where supposed to bring :) (enhanced discoverability, better readability)

**Comment by mateusfccp** [comment]
> Coming back to this:
> 
> For quite some time now, the community has started using extension methods as a workaround to the lack of overloading.

Wow, that's disgusting. I've never seen it in the wild, and I had no idea people were doing it.

**Comment by FMorschel** [comment]
Maybe we could start with a temporary solution like what I proposed at https://github.com/dart-lang/sdk/issues/60693.

The idea is a new meta annotation where we can enforce specific types. So we could do:

```dart
@AcceptedTypes([List, Stream, Future])
void fn(Object o) {...}
```

Or something like that. The idea is not fully developed yet. This came from Flutter changing the types for some `ThemeData` proprieties recently. They were using `Object` and if the passed type was not one of the supported types, the user would only find out when running.

This is more of a solution to union types, but maybe could be applied here too. If you have any suggestions regarding that, please say so in the above-linked issue.

**Comment by TekExplorer** [comment]
Union types are great, (and its own issue) but one of the benefits of overloading is that extensions could add to the overloads, expanding support. This could result in union-like support, but instead of getting a union type as a value, you get a known type based on the parameters. My concern actually, is that it could get weird if you have 2 valid options if your type implements two types that have their own overloads.

Perhaps we could make labels more useful here? we could require that every overload must have a unique label, and if they collide across extensions, you'll need to select the extension as normal.

Granted, im not sure how exactly you'd select it, but i'm sure we can argue syntax later

huh. that might be all we need. thats not to say its _simple_ of course, but i think its worth it.

```dart
class Foo {
  unary:
  void thing() {}
  
  string:
  String thing(String str) => str;

  list:
  List thing(List list) => list;
  
  map:
  Map thing(Map map) => map;
}

class ML implements List, Map {...}

check(ML weird) {
  final foo = Foo();
  foo.thing();
  foo.thing('...');
  foo.thing([]);
  foo.thing({});
  foo.thing(weird); // possibly errors; ambiguous overload

  SomeExtension(foo).thing(otherType);

  foo.thing:list(weird);
  foo.thing::list(weird);
  foo.thing:list:(weird);
  // some other syntax? should be relatively rare
  // i imagine that we could alternaitively go for a "first-fit" option,
  // or have a language option to require selection if its ambiguous
}
```
or something.

**Comment by rrousselGit** [comment]
Unions don't solve overloading.

Overloading include things like parameter combination. For example, you could do:

```dart
forEach(void Function(T value));
forEach(void Function(T value, int index));
```


Or you can even merge all the `Padding` constructors in one:

```dart
Padding({int all})
Padding({int left, int right, int top, int bottom})
Padding({int horizontal, int vertical})
```

The constructors are mutually exclusive. If you do `Padding(left: 0)`, you cannot also specify `horizontal`.

Union types on their own wouldn't help. Well, at least not at the parameter level. Overloading is more similar to:

```dart
Padding Function({int all})
  | Padding Function({int left, int right, int top, int bottom})
  | Padding Function({int horizontal, int vertical})
 constructor;
```

But where you can do `constructor(all: ...)` directly instead of needing `if (constructor is Function({int all)) constructor(all: 42)`

**Comment by rrousselGit** [comment]
Btw about this:

> > Coming back to this:
> > For quite some time now, the community has started using extension methods as a workaround to the lack of overloading.
> 
> Wow, that's disgusting. I've never seen it in the wild, and I had no idea people were doing it.

Using extension types for pseudo method overloading is something Dart itself does:

```dart
final aFuture = Future.value(42);
final bFuture = Future.value('Hello');
...
final (a, b) = await (aFuture, bFuture).wait;
```

It uses extensions to support `(a,).wait`, `(a, b).wait`, `(a, b, c).wait`, ...


If the Dart SDK itself uses this trick, it's fair to say that this is a useful thing.

**Comment by talski** [comment]
```dart
import 'package:flutter/material.dart';

class MyWidget<T> {
  final Widget Function(T item, [DateTime timestamp]) itemBuilder;
  final T current;

  MyWidget({required this.current, required this.itemBuilder});

  build(BuildContext context) {
    if (DateTime.now().hour < 8) {
      return itemBuilder(current); // I can call it with only first param
    }
    
    return itemBuilder(current, DateTime.now());
  }
}

void main() {
  final first = MyWidget(
    current: 'My First Value',
    itemBuilder: (item, [DateTime? timestamp]) => Text('$item$timestamp'),
  );

  final second = MyWidget(
    current: 'My Second Value',
    itemBuilder: (item) => Text(item), // I can't declare a function omitting an optional param
  );
}
```

What is the reason for this not to work? timestamp is an optional param, any news on this issue?

**Comment by FMorschel** [comment]
@talski, for your problem specifically, you can make the closure like `(item, [_]) => Text(item)`.

**Comment by talski** [comment]
Sure, but I will need to change the closure everywhere I use this widget, would be nice to keep it as is, and only change where I need this new parameter, as it is optional

**Comment by FMorschel** [comment]
Yes, I agree. While this is not possible, you might be interested in the following:

- https://github.com/dart-lang/sdk/issues/46972

**Comment by lrhn** [comment]
It doesn't work because if it did, the type system would not be sound.

The type `Widget Function(T item, [DateTime timestamp])` it's the type of a function that can be called with both one and two arguments. 
Code can take one such a value and call it with one argument, and then with two arguments.


A function that can only be called with one argument (or only with two arguments) cannot have that type. Because if it did, someone could call that function with two (or one) argument, which it doesn't support at all.

(Also not related to the issue.)

**Comment by rrousselGit** [comment]
> It doesn't work because if it did, the type system would not be sound.

Does this not depend on how it is specified?

Couldn't we do something specific to closure, where when a closure is created, we look at the context type and implicitly add all the omitted parameters?

Such that we have

```dart
// Allowed
void Function(int) cb = () {};
// This is strictly identical to:
void Function(int) cb = (_) {};


// And as such we have:
print(cb is void Function()); // false
print(cb is void Function(int)); // true

// No 'downcast'/'upcast' either
void Function() a;
void Function(int) b = a; // ERROR, not assignable

// We could provide a small sugar to avoid having to do `() => a()`
// (although such feature is very much optional)
void Function() a;
void Function(int) b = a.bind;
// ^OK, identical to:
void Function(int) b = (_) => a().
```

So there would be not function inheritance or anything of the sort. It'd just be sugar for not having to write unused parameters in closures


(still not related to the issue though, sorry)
